### PR TITLE
Precompiled scripts

### PR DESCRIPTION
**Pull request by aslesarenko**
In this PR:
- updates sigma dependency and modifies ErgoInterpreter so that it uses new PrecompileScriptProcessor to more efficiently execute complex scripts.
- It also defines a list of pre-defined top 10 scripts by number of usages
- Simple P2PK scripts are executed directly without building any graphs in IRContext which makes reduction to crypto very fast.

### CODE REVIEW

File: ergo-wallet/src/main/scala/org/ergoplatform/wallet/interpreter/ErgoInterpreter.scala [hunk 1]
```diff
@@ -82,12 +87,62 @@ class ErgoInterpreter(params: ErgoLikeParameters)(implicit IR: IRContext)
...
+    */
+  val scriptProcessor = {
+    val scriptKeys = frequentlyUsedScripts.flatMap { s =>
+      val bytes = ErgoAlgos.decodeUnsafe(s)
+      validationSettings.map(vs => CacheKey(bytes, vs))
+    }
+    new PrecompiledScriptProcessor(
+      ScriptProcessorSettings(
+        predefScripts = scriptKeys,
+        maxCacheSize = 2000,  // takes up around 600Mb
```
Comment (kushti) [inline ergo-wallet/src/main/scala/org/ergoplatform/wallet/interpreter/ErgoInterpreter.scala:137]: We have nodes running on Raspberry Pi with 512 MB limit, 600 Mb cache is too big

File: ergo-wallet/src/main/scala/org/ergoplatform/wallet/interpreter/ErgoInterpreter.scala [hunk 2]
```diff
@@ -82,12 +87,62 @@ class ErgoInterpreter(params: ErgoLikeParameters)(implicit IR: IRContext)
...
+        recordCacheStats = true,
+        reportingInterval = 500
+      )) with ScorexLogging {
+      override protected def createIR(): IRContext = new RuntimeIRContext
+
+      override protected def onReportStats(stats: ProcessorStats): Unit = {
+        log.warn(s"Stats: $stats")
```
Comment (kushti) [inline ergo-wallet/src/main/scala/org/ergoplatform/wallet/interpreter/ErgoInterpreter.scala:144]: Why warn level?

File: ergo-wallet/src/main/scala/org/ergoplatform/wallet/interpreter/ErgoInterpreter.scala [hunk 3]
```diff
@@ -82,12 +87,62 @@ class ErgoInterpreter(params: ErgoLikeParameters)(implicit IR: IRContext)
 
 object ErgoInterpreter {
 
+  /** Creates an interpreter with the given parameters, [[RuntimeIRContext]] for script processing and
+    * the `scriptProcessor` for handling compilation caching and handling pre-compiled scripts.
+    */
   def apply(params: ErgoLikeParameters): ErgoInterpreter =
-    new ErgoInterpreter(params)(new RuntimeIRContext)
+    new ErgoInterpreter(params)(new RuntimeIRContext) {
+      override val precompiledScriptProcessor = scriptProcessor
```
Comment (kushti) [inline ergo-wallet/src/main/scala/org/ergoplatform/wallet/interpreter/ErgoInterpreter.scala:95]: Type annotation is needed for public member

File: ergo-wallet/src/main/scala/org/ergoplatform/wallet/interpreter/ErgoInterpreter.scala [hunk 4]
```diff
@@ -1,13 +1,18 @@
 package org.ergoplatform.wallet.interpreter
 
+import com.google.common.cache.CacheStats
```
Comment (kushti) [inline ergo-wallet/src/main/scala/org/ergoplatform/wallet/interpreter/ErgoInterpreter.scala:3]: Unused import added

File: ergo-wallet/src/main/scala/org/ergoplatform/wallet/interpreter/ErgoInterpreter.scala [hunk 5]
```diff
@@ -82,12 +87,62 @@ class ErgoInterpreter(params: ErgoLikeParameters)(implicit IR: IRContext)
...
+    /*14435*/"100c0e20dcd3cdd11102e7cb675fd5185a1685ad007619badcf398864274b137d9f45a9e0e2019323d02836f596329b28b2d380d3cf27a10f55db7b1d33bc4537ab538594bae0402040005c08db701040404000404040205c08db70104060400d805d601d901010493cbc2b2a47201007300d602c5a7d603c2a7d604e4c6a70407d6057301eb02d1ecedededda7201017302dad901060493c5b2a47206007202017303dad9010604d801d608b2a5720600eded93c27208720393c1720899c1a7730493e4c6720804077204017305dad901066393cbc27206720501b2a5730600edededda7201017307dad901060493c5b2a47206007202017308dad9010604d801d608b2a5720600eded93c27208720393c1720899c1a7730993e4c672080407720401730adad901066393cbc27206720501b2a4730b00cd7204"
+  )
+
+  /** Script processor which uses [[RuntimeIRContext]] to process graphs.
+    * Preforms pre-compilation of the given scripts during instantiation.
+    * Keeps pre-compiled data structures for the lifetime of JVM.
+    */
+  val scriptProcessor = {
+    val scriptKeys = frequentlyUsedScripts.flatMap { s =>
```
Comment (kushti) [inline ergo-wallet/src/main/scala/org/ergoplatform/wallet/interpreter/ErgoInterpreter.scala:130]: Please comment logic for script keys

File: ergo-wallet/src/main/scala/org/ergoplatform/wallet/interpreter/ErgoInterpreter.scala [hunk 6]
```diff
@@ -82,12 +87,62 @@ class ErgoInterpreter(params: ErgoLikeParameters)(implicit IR: IRContext)
...
+      override val precompiledScriptProcessor = scriptProcessor
+    }
 
+  /** Create [[AvlTreeData]] with the given digest and all operations enabled. */
   def avlTreeFromDigest(digest: ADDigest): AvlTreeData = {
     val flags = AvlTreeFlags(insertAllowed = true, updateAllowed = true, removeAllowed = true)
     AvlTreeData(digest, flags, Constants.HashLength)
   }
 
+  /** A list of settings, one for each soft-fork. */
```
Comment (kushti) [inline ergo-wallet/src/main/scala/org/ergoplatform/wallet/interpreter/ErgoInterpreter.scala:104]: The ScalaDoc comment is not descriptive. Please provide more details.

**Review by kushti — CHANGES_REQUESTED** [review]

File: ergo-wallet/src/main/scala/org/ergoplatform/wallet/interpreter/ErgoInterpreter.scala (hunk 1 above)
Comment (aslesarenko) [reply ergo-wallet/src/main/scala/org/ergoplatform/wallet/interpreter/ErgoInterpreter.scala:137]: reduced to 500

File: ergo-wallet/src/main/scala/org/ergoplatform/wallet/interpreter/ErgoInterpreter.scala (hunk 2 above)
Comment (aslesarenko) [reply ergo-wallet/src/main/scala/org/ergoplatform/wallet/interpreter/ErgoInterpreter.scala:144]: changed to `info`.

File: ergo-wallet/src/main/scala/org/ergoplatform/wallet/interpreter/ErgoInterpreter.scala (hunk 3 above)
Comment (aslesarenko) [reply ergo-wallet/src/main/scala/org/ergoplatform/wallet/interpreter/ErgoInterpreter.scala:95]: fixed

File: ergo-wallet/src/main/scala/org/ergoplatform/wallet/interpreter/ErgoInterpreter.scala (hunk 4 above)
Comment (aslesarenko) [reply ergo-wallet/src/main/scala/org/ergoplatform/wallet/interpreter/ErgoInterpreter.scala:3]: fixed

File: ergo-wallet/src/main/scala/org/ergoplatform/wallet/interpreter/ErgoInterpreter.scala (hunk 5 above)
Comment (aslesarenko) [reply ergo-wallet/src/main/scala/org/ergoplatform/wallet/interpreter/ErgoInterpreter.scala:130]: simplified and commented

File: ergo-wallet/src/main/scala/org/ergoplatform/wallet/interpreter/ErgoInterpreter.scala (hunk 6 above)
Comment (aslesarenko) [reply ergo-wallet/src/main/scala/org/ergoplatform/wallet/interpreter/ErgoInterpreter.scala:104]: This is actually not necessary, removed.

**Comment by aslesarenko** [comment]
Found one issue with cost accumulation, need to fix it on sigma side first.

**Review by kushti — APPROVED** [review]

**Pull request merged**